### PR TITLE
AMQP-641: DirectMLContainer Add ErrorHandler

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-641

Simply change `invokeListener` to `executeListener` which encapsulates
error handling (and message post processing and de-batching logic).

### CODE REVIEW

**Comment by artembilan** [comment]
H-m.
We still have some race condition...:
```
org.springframework.amqp.rabbit.listener.DirectMessageListenerContainerTests > testQueueManagement FAILED

    java.lang.AssertionError at DirectMessageListenerContainerTests.java:168
```

**Comment by garyrussell** [comment]
Yes; I noticed that (locally) yesterday - will push a fix.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java [hunk 1]
```diff
@@ -515,7 +507,9 @@ private void removeConsumer() {
 				DirectMessageListenerContainer.this.cancellationLock.release(this);
 				DirectMessageListenerContainer.this.consumers.remove(this);
 			}
+			RabbitUtils.setPhysicalCloseRequired(true);
 			RabbitUtils.closeChannel(getChannel());
+			RabbitUtils.setPhysicalCloseRequired(false);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: What made you to do this?
Looks like `isPhysicalCloseRequired()` takes care about `ThreadLocal` cleanup:
```
public static boolean isPhysicalCloseRequired() {
		Boolean mustClose = physicalCloseRequired.get();
		if (mustClose == null) {
			mustClose = Boolean.FALSE;
		}
		else {
			physicalCloseRequired.remove();
		}
		return mustClose;
}
```

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerTests.java [hunk 2]
```diff
@@ -235,6 +248,34 @@ public void publishEvent(ApplicationEvent event) {
...
+				.get());
+		brokerRunning.getAdmin().declareQueue(q1);
+		CachingConnectionFactory cf = new CachingConnectionFactory("localhost");
+		DirectMessageListenerContainer container = new DirectMessageListenerContainer(cf);
+		container.setQueueNames(Q1);
+		container.setConsumersPerQueue(2);
+		final AtomicReference<Channel> channel = new AtomicReference<>();
+		container.setMessageListener((ChannelAwareMessageListener) (m, c) -> {
+			channel.set(c);
+			throw new MessageConversionException("intended - should be wrapped in an ARADRE");
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerTests.java]: I'd prefer the full exception name: `AmqpRejectAndDontRequeueException`.
We don't chat here in Slack :smile:

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java (hunk 1 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: Oops - good catch.

File: spring-amqp/src/main/java/org/springframework/amqp/utils/MapBuilder.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Map;
+
+/**
+ * A {@code Builder} pattern implementation for a {@link Map}.
+ *
+ * @author Artem Bilan
+ * @author Gary Russell
+ * @since 2.0
+ */
+public class MapBuilder<B extends MapBuilder<B, K, V>, K, V> {
```
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/utils/MapBuilder.java]: Well, you don't have JavaDocs for generic variable and neither `gredlew api`, nor `gredlew check` barks.
So, in which cases do we have problem with JavaDocs on the matter?

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerTests.java [hunk 4]
```diff
@@ -235,6 +248,34 @@ public void publishEvent(ApplicationEvent event) {
...
+		});
+		container.afterPropertiesSet();
+		container.start();
+		RabbitTemplate template = new RabbitTemplate(cf);
+		template.convertAndSend(Q1, "foo");
+		assertNotNull(template.receive(DLQ1, 10000));
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerTests.java]: OK. That took some time for me to figure out that `ConditionalRejectingErrorHandler` is used by default and what it does exactly.
But, eh..., right now it's fine with me.
Pretty smart and cool.

I can merge it with fixes for those minor concerns.
Just let me know!

Comment (garyrussell) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerTests.java]: Thanks; sounds good - I have the transaction support waiting for this merge 😄

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/1008009bd0bd50d3d048159d6ba531b76e1bcf6c

**Pull request closed without merging**